### PR TITLE
[Core] trim size of Reference struct

### PR DESCRIPTION
**Pull request by mwtian**
## Why are these changes needed?
During large scale shuffle (number of partitions used >= 1000), driver uses significant amount of memory for storing ObjectRefs. On Intel MacOS, each `Reference` struct currently takes up 592 bytes. We can reduce per-`Reference` memory footprint:
- During shuffle, no ObjectRef borrowing or nesting happens. And in this case fields related to borrowing or nesting should not take up memory. This reduces `sizeof(Reference)` from 592 to 400.
- Fields in the `Reference` struct can be reordered to enhance packing. This reduces `sizeof(Reference)` from 400 to 368.

On Intel MacOS running the shuffle benchmark with 1000 partitions and 10MB partition size, RSS at the end of shuffle drops from ~5GB to ~4.5GB.

## Related issue number
#23604

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by mwtian** [comment]
On Ubuntu 20.04 with tcmalloc, I see similar ~10% reduction in final RSS memory usage from ~6.5GB to ~5.9GB. Heap profile from tcmalloc:
image

File: src/ray/core_worker/reference_count.h [hunk 1]
```diff
@@ -587,89 +664,71 @@ class ReferenceCounter : public ReferenceCounterInterface,
...
+
+    /// Metadata for references that contain this reference.
+    std::unique_ptr<ContainingReferences> containing_references;
+
+    /// Metadata related to borrowing.
+    std::unique_ptr<BorrowInfo> borrow_info;
+
+    /// Callback that will be called when this ObjectID no longer has
+    /// references.
+    std::function<void(const ObjectID &)> on_delete;
```
Comment (stephanie-wang) [inline src/ray/core_worker/reference_count.h:698]: We can do it in a different PR, but out of curiosity, did you try testing what would happen if we use the same callback for all objects instead of saving them individually?

**Review by stephanie-wang — COMMENTED** [review]
Looks good! A couple questions about testing:
1. Were you able to confirm that the shuffle benchmark doesn't set the new optional fields?
2. Could we add this to the microbenchmark suite and start tracking memory usage in the perf dashboard?

File: src/ray/core_worker/reference_count.h [hunk 2]
```diff
@@ -563,21 +610,51 @@ class ReferenceCounter : public ReferenceCounterInterface,
       }
     }
 
+    /// Access BorrowInfo without modifications.
+    /// Returns the default value of the struct if it is not set.
+    const BorrowInfo &borrow() const {
+      if (borrow_info == nullptr) {
+        static auto *default_info = new BorrowInfo();
```
[the remark this answers is not in the document]
Comment (mwtian) [reply src/ray/core_worker/reference_count.h:617]: Thanks for the suggestion. Using `static local` raw pointers to avoid the objects being destructed was a common idiom in my previous company, so people need not think about the implication of destroying `default_info` on shutdown. But I can update this if it conflicts with Ray's style.

**Comment by mwtian** [comment]
> Looks good! A couple questions about testing:
> 
> 1. Were you able to confirm that the shuffle benchmark doesn't set the new optional fields?
> 2. Could we add this to the microbenchmark suite and start tracking memory usage in the perf dashboard?

Yes the new optional fields do not show up in the profile. I can add some check failure too. I can add a per-object ref memory benchmark separately.

File: src/ray/core_worker/reference_count.h [hunk 3]
```diff
@@ -494,6 +494,50 @@ class ReferenceCounter : public ReferenceCounterInterface,
   void ReleaseAllLocalReferences();
 
  private:
+  struct ContainingReferences {
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply src/ray/core_worker/reference_count.h:497]: I think the fields actually are counts if you take the size of the sets but yeah we can use `NestedReferenceInfo` or something, maybe?

[the remark this answers is not in the document]
Comment (mwtian) [reply src/ray/core_worker/reference_count.h]: Sounds good, renamed to `NestedReferenceCount`.

**Review by stephanie-wang — APPROVED** [review]
